Add CWL support (resolves #403)

tetron opened the pull request:
Prototype CWL support for Toil from BioHackathon 2015.

cgcloud-jenkins commented:
Can one of the admins verify this patch?

benedictpaten commented:
Jenkins, please test

hannes-ucsc commented:
Jenkins, test this please.

hannes-ucsc commented:
Can this be rebased on master or does it have to be based on @benedictpaten's PR #397?

tetron commented:
You'll have to talk to @benedictpaten, he advised me to base it on his branch, I believe it may rely on API changes specific to that branch.

cgcloud-jenkins commented:
We'll do the rebase once we get the parent branch in.

On Fri, Sep 18, 2015 at 2:17 PM, Peter Amstutz <notifications@github.com>
wrote:

> You'll have to talk to @benedictpaten <https://github.com/benedictpaten>,
> he advised me to base it on his branch, I believe it may rely on API
> changes specific to that branch.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BD2KGenomics/toil/pull/404#issuecomment-141347488>.
>
> --
> You received this message because you are subscribed to the Google Groups
> "cgloud-jenkins" group.
> To unsubscribe from this group and stop receiving emails from it, send an
> email to cgloud-jenkins+unsubscribe@googlegroups.com.
> To post to this group, send email to cgloud-jenkins@googlegroups.com.
> To view this discussion on the web visit
> https://groups.google.com/d/msgid/cgloud-jenkins/BD2KGenomics/toil/pull/404/c141347488%40github.com
> <https://groups.google.com/d/msgid/cgloud-jenkins/BD2KGenomics/toil/pull/404/c141347488%40github.com?utm_medium=email&utm_source=footer>
> .
>
> For more options, visit https://groups.google.com/d/optout.
>

hannes-ucsc commented:
The branch this PR's branch was based on has been merged into master. Please rebase this  PR's branch onto master.

tetron commented:
Rebased.  Also updated setup.py.  See code for comments about remaining work tasks.  Let's talk about how to proceed, @benedictpaten suggested that someone from the toil team would be able to take this branch and finish it.

benedictpaten commented:
Yep, we'll discuss at our planning meeting. Thanks! Awesome to chat last
week, and look forward to chatting again soon.

On Mon, Sep 21, 2015 at 2:49 PM, Peter Amstutz <notifications@github.com>
wrote:

> Rebased. Also updated setup.py. See code for comments about remaining work
> tasks. Let's talk about how to proceed, @benedictpaten
> <https://github.com/benedictpaten> suggested that someone from the toil
> team would be able to take this branch and finish it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BD2KGenomics/toil/pull/404#issuecomment-142120325>.
>

hannes-ucsc commented on Makefile:
```diff
@@ -21,9 +21,9 @@ Supported targets: 'develop', 'docs', 'sdist', 'clean', 'test', 'pypi', or 'pypi
-extras. For example, to install Toil in develop mode with Mesos, AWS and Azure support, run
+extras. For example, to install Toil in develop mode with Mesos, AWS, Azure and CWL support, run
 
-make develop extras=[mesos,aws,azure]
```
The README now mentions the extras. I'm inclined to remove this example from the Makefile as it would be yet another place to modify if you add an extra to Toil.

tetron commented:
You can now test cwltoil using the CWL conformance tests:

```
cd common-workflow-language/conformance
./run_test.sh   TOIL=$HOME/work/toil CWLTOOL=""
Test [1/36]
...
Test [36/36]
13 failures
```

(Still needs some work)

benedictpaten commented:
Cool! Thanks!

On Wed, Sep 23, 2015 at 8:36 AM, Peter Amstutz <notifications@github.com>
wrote:

> You can now test cwltoil using the CWL conformance tests:
>
> cd common-workflow-language/conformance
> ./run_test.sh   TOIL=$HOME/work/toil CWLTOOL=""
> Test [1/36]
> ...
> Test [36/36]
> 13 failures
>
> (Still needs some work)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BD2KGenomics/toil/pull/404#issuecomment-142641710>.
>

tetron commented:
I'm continuing to update this, with the goal of getting it to the point where it passes all conformance tests.  I've only tested with local host runner, and there are some details around file staging we should discuss.  However, let's talk about what your expectations are for this to be ready for code review and merge?

hannes-ucsc commented:
I think the only thing that is missing is a unit test. I see a test directory in this patch but no a test code. If the test code uses single machine that would be sufficient. I can then go ahead and extend the test to use the other batch systems. I'm setting the `needs work` label. Whenever you think this PR is done from your point of view, remove the label and I take it from there.

hannes-ucsc commented:
Oops, label as already set. Even better.

tetron commented:
Hi @benedictpaten, @hannes-ucsc 

I think this PR is ready for you to take over.  I have rebased on master, "make test" passes, and it passes 30/36 CWL conformance tests, with the remaining 6 tests due to unimplemented features (these are left as an exercise to the reader).

For a batch deployment, some additional work is probably required for file staging.  The current solution is to use a "staging" job to put local files into the file store, however this assumes that the staging job runs on the head node.  It also need to be tweaked to recognize files that are already in cloud storage (such as S3 buckets).

I'm happy to make tweaks if you have specific review comments that you require for merge, otherwise I don't plan on doing any more work on this branch, so please take it and run with it.

hannes-ucsc commented:
Jenkins, ok to test.

hannes-ucsc commented:
Peter (@tetron), thank you very much for this contribution! I will take it from here and only bother you if I have specific questions.

benedictpaten commented:
Thanks Peter! This isn't part of our current sprint, but we'll definitely
try and get this into the next one. It will be great to have CWL support! I
agree the staging issue needs to get sorted out.

On Tue, Sep 29, 2015 at 12:22 PM, Peter Amstutz <notifications@github.com>
wrote:

> Hi @benedictpaten <https://github.com/benedictpaten>, @hannes-ucsc
> <https://github.com/hannes-ucsc>
>
> I think this PR is ready for you to take over. I have rebased on master,
> "make test" passes, and it passes 30/36 CWL conformance tests, with the
> remaining 6 tests due to unimplemented features (these are left as an
> exercise to the reader).
>
> For a batch deployment, some additional work is probably required for file
> staging. The current solution is to use a "staging" job to put local files
> into the file store, however this assumes that the staging job runs on the
> head node. It also need to be tweaked to recognize files that are already
> in cloud storage (such as S3 buckets).
>
> I'm happy to make tweaks if you have specific review comments that you
> require for merge, otherwise I don't plan on doing any more work on this
> branch, so please take it and run with it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BD2KGenomics/toil/pull/404#issuecomment-144161799>.
>

hannes-ucsc commented:
to be continued in PR #448 ...
